### PR TITLE
testcase/le_tc/kernel : Add Testcases for workqueue

### PR DESCRIPTION
**Pull request by kavyasrinunna**
Covers apis :
work_cancel()
work_queue()

Signed-off-by: Kavya <kavya.nunna@partner.samsung.com>

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check Result:
apps/examples/testcase/le_tc/kernel/tc_wqueue.c:143: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxV)

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+	TC_ASSERT_EQ_CLEANUP("work_cancel", result, OK, goto cleanup);
+
+	result = work_queue(LPWORK, test_work3, wq_test3, (void *)0, 10);
+	TC_ASSERT_EQ_CLEANUP("work_queue",result, OK, goto cleanup);
+
+	sleep(1);
+
+
+cleanup:
+		free(test_work1);
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: check the alignment

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+#include <fcntl.h>
+#include "tc_internal.h"
+
+/**************************************************************************
+* Private Definitions
+**************************************************************************/
+
+
+#if defined(CONFIG_SCHED_WORKQUEUE) || defined(CONFIG_LIB_USRWORK)
+
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: remove unnecessary lines

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+/**************************************************************************
+* Included Files
+**************************************************************************/
+#include <tinyara/config.h>
+#include <tinyara/wqueue.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <ctype.h>
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Is this header needed?

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	result = work_cancel(LPWORK, test_work2);
+	TC_ASSERT_EQ_CLEANUP("work_cancel", result, OK, goto cleanup);
+
+	result = work_queue(HPWORK, test_work3, wq_test3, (void *)0, 10);
+	TC_ASSERT_EQ_CLEANUP("work_queue", result, OK, goto cleanup);
+	result = work_cancel(HPWORK, test_work3);
+	TC_ASSERT_EQ_CLEANUP("work_cancel", result, OK, goto cleanup);
+
+	result = work_queue(LPWORK, test_work3, wq_test3, (void *)0, 10);
+	TC_ASSERT_EQ_CLEANUP("work_queue",result, OK, goto cleanup);
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: space is needed between params

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+**************************************************************************/
+
+/**************************************************************************
+* Included Files
+**************************************************************************/
+#include <tinyara/config.h>
+#include <tinyara/wqueue.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Is this header needed?

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+*
+**************************************************************************/
+
+/**************************************************************************
+* Included Files
+**************************************************************************/
+#include <tinyara/config.h>
+#include <tinyara/wqueue.h>
+#include <stdio.h>
+#include <string.h>
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Is this header needed?

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+	int result;
+	struct work_s *test_work1;
+	struct work_s *test_work2;
+	struct work_s *test_work3;
+
+	test_work1 = (struct work_s *)malloc(sizeof(struct work_s));
+	test_work2 = (struct work_s *)malloc(sizeof(struct work_s));
+	test_work3 = (struct work_s *)malloc(sizeof(struct work_s));
+
+	start_time = clock_systimer();
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: for clock_systimer, please add the proper header file

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+* Included Files
+**************************************************************************/
+#include <tinyara/config.h>
+#include <tinyara/wqueue.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <ctype.h>
+#include <fcntl.h>
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Is this header needed?

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 3 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Will remove it, It is not needed

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 2 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: okay

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 1 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: okay

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 6 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: will remove it

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 5 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: will remove it

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 8 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: will remove it

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 7 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: will add it

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 4 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: okay

**Comment by kavyasrinunna** [comment]
Modified as per review comments

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/testcase/le_tc/kernel/Kconfig [hunk 9]
```diff
@@ -207,7 +208,9 @@ config TC_KERNEL_TASH_HEAPINFO
 config TC_KERNEL_TASH_STACKMONITOR
 	bool "StackMonitor"
 	default n
-
+config TC_KERNEL_WORK_QUEUE
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/Kconfig]: Could you make a empty line before and after to sparate different config?

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/// @file wqueue.c
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: tc_wqueue.c

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+/// @brief Test Case Example for Work Queue API
+/**************************************************************************
+*
+*   Copyright (C) 2007-2009, 2011 Gregory Nutt. All rights reserved.
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Let's remove Nuttx license from this file.

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+{
+	systime_t cur_time = 0;
+	cur_time = clock_systimer();
+	printf("workqueue_test 3 : test 3 requested delay is (%u) ticks, executed delay is (%llu) ticks.\n", (uint32_t)arg, (uint64_t)cur_time - (uint64_t)start_time);
+}
+/**************************************************************************
+* Public Functions
+**************************************************************************/
+#if defined(CONFIG_SCHED_HPWORK) || defined(CONFIG_SCHED_LPWORK)
+static void tc_wqueue_cancel(void)
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Why is function name (tc name) cancel? only cancel is main content? Only cancel without any normal operation of wqueue doesn't make sense.

Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Okay , Can I name it as tc_wqueue_create_cancel ??

File: apps/examples/testcase/le_tc/kernel/Kconfig (hunk 9 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/Kconfig]: Okay

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 11 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Done

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 10 above)
Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: okay

File: apps/examples/testcase/le_tc/kernel/tc_wqueue.c (hunk 12 above)
Comment (sunghan-chang) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Our guidance is adding function names so that ```tc_wqueue_work_queue_cancel``` is correct.

Comment (kavyasrinunna) [reply apps/examples/testcase/le_tc/kernel/tc_wqueue.c]: Okay

**Comment by kavyasrinunna** [comment]
Modified as per review comments

**Comment by seinfra** [comment]
Code Rule Check OK.

**Pull request merged**